text/0000-bench-utils.md

```diff
@@ -0,0 +1,207 @@
+- Feature Name: black_box
+- Start Date: 2018-03-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds one function, `core::hint::black_box`, which is a hint to the
+optimizer to disable certain compiler optimizations. 
+
+# Motivation
+[motivation]: #motivation
+
+A tool for preventing compiler optimizations is widely useful. One application
+is writing synthetic benchmarks, where, due to the constrained nature of the
+benchmark, the compiler is able to perform optimizations that wouldn't otherwise
+trigger in practice. Another application is writing constant time code, where it
+is undesirable for the compiler to optimize certain operations depending on the
+context in which they are executed.
+
+The implementation of this function is backend-specific and currently requires
+inline assembly. No viable alternative is available in stable Rust.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `hint::black_box`
+
+The function:
+
+```rust
+/// An _unknown_ function that returns `x`.
+pub unsafe fn black_box<T>(x: T) -> T;
+```
+
+is an _unknown_ function, that is, a function that the compiler cannot make any
+assumptions about. It can potentially use `x` in any possible valid way that
+`unsafe` Rust code is allowed to, and requires the compiler to be maximally
+pessimistic in terms of optimizations. The compiler is still allowed to optimize
+the expression generating `x`. This function returns `x` and is a no-op in the
+virtual machine.
```

Comment:
Which virtual machine?

I know what you mean but many readers will likely be confused.

Reply:
Yeah I don't think there is a good way to put it right now beyond maybe saying "abstract" instead of "virtual".

If we had a memory model, that memory model would specify the abstract machine that Rust runs on and on which the memory model is valid, and I could just refer to _that_ abstract machine here. 

Maybe I can rephrase that as "is a no-op in the abstract machine of Rust's memory model (whatever that memory model might end up being)" or something like that ?

Reply:
I don't think most people will know what this means.

I would say something like "*You* can rely on `black_box` being a NOP just returning `x`, but the compiler will optimize under the pessimistic assumption that `black_box` might do anything with the data it got".